Telegram voice messages bypass the bundled STT module — empty text reaches Claude and breaks session resume

## Telegram voice messages bypass the bundled STT module — empty `text` reaches Claude and breaks session resume

### Summary

The repo already ships a `stt` module (`packages/jimmy/src/stt/stt.ts`) plus a portal flow to download Whisper models, and a working `stt.enabled / model / languages` config block. However, the Telegram connector (`packages/jimmy/src/connectors/telegram/index.ts`) only forwards `telegramMsg.text` and never invokes STT. When a user sends a voice/audio/video-note message, the connector emits an `IncomingMessage` with `text: ""`, which then crashes Claude session resume with:

```
Error: No deferred tool marker found in the resumed session.
Either the session was not deferred, the marker is stale (tool already ran),
or it exceeds the tail-scan window. Provide a prompt to continue the conversation.
```

### Reproduction

1. Configure a Telegram connector and enable STT in `~/.jinn/config.yaml`:
   ```yaml
   connectors:
     telegram:
       botToken: ...
       allowFrom: [...]
   stt:
     enabled: true
     model: small
     languages: [en, pt]
   ```
2. Use the portal to download the Whisper `small` model (works — file lands in `~/.jinn/models/whisper/ggml-small.bin`).
3. Install runtime deps the STT module shells out to: `whisper-cli` (whisper.cpp) and `ffmpeg` on PATH.
4. Send a voice message to the bot.
5. The connector forwards an empty-text message; gateway logs show:
   ```
   [ERROR] Claude exited with code 1: Error: No deferred tool marker found in the resumed session.
   [WARN]  Dead session detected for <id>: ...
   ```

### Root cause

`packages/jimmy/src/connectors/telegram/index.ts` only reads `telegramMsg.text`. There is no detection of `voice`, `audio`, or `video_note`, no call into the existing `stt/stt.ts` module, no file download from the Telegram Bot API.

### Proposed fix

Wire the existing `transcribe()` function into the Telegram message handler. Two small changes:

**1. `packages/jimmy/src/shared/types.ts`** — extend `TelegramConnectorConfig`:

```ts
export interface TelegramConnectorConfig {
  id?: string;
  employee?: string;
  botToken: string;
  allowFrom?: number[];
  ignoreOldMessagesOnBoot?: boolean;
  /** STT settings (forwarded from top-level `config.stt`) */
  stt?: GatewayConfig["stt"];
}
```

**2. `packages/jimmy/src/gateway/server.ts`** — pass `config.stt` into `TelegramConnector` at each construction site (3 today: the singleton telegram block, plus the two named-instance branches):

```ts
const telegram = new TelegramConnector({
  botToken: config.connectors.telegram.botToken,
  allowFrom: config.connectors.telegram.allowFrom,
  ignoreOldMessagesOnBoot: config.connectors.telegram.ignoreOldMessagesOnBoot,
  stt: config.stt,
});
```

**3. `packages/jimmy/src/connectors/telegram/index.ts`** — detect voice-like messages and transcribe before forwarding:

```ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import {
  transcribe as sttTranscribe,
  resolveLanguages,
  getModelPath,
} from "../../stt/stt.js";

// in constructor:
private readonly sttConfig?: TelegramConnectorConfig["stt"];

constructor(config: TelegramConnectorConfig) {
  // ...existing assignments...
  this.sttConfig = config.stt;
}

// inside the bot.on("message", ...) handler, before building `msg`:
let messageText = telegramMsg.text || telegramMsg.caption || "";

const voiceLike =
  telegramMsg.voice ?? telegramMsg.audio ?? telegramMsg.video_note;

if (voiceLike && this.sttConfig?.enabled) {
  try {
    const model = this.sttConfig.model || "small";
    if (!getModelPath(model)) {
      logger.warn(`[telegram] STT model '${model}' not downloaded — skipping`);
    } else {
      const langs = resolveLanguages(this.sttConfig);
      const language = langs.length === 1 ? langs[0] : "auto";
      logger.info(
        `[telegram] Transcribing voice message (${voiceLike.duration}s, lang=${language})`,
      );
      try { await this.bot.sendChatAction(telegramMsg.chat.id, "typing"); } catch {}

      const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "tg-stt-"));
      try {
        const localPath = await this.bot.downloadFile(voiceLike.file_id, tmpDir);
        const text = await sttTranscribe(localPath, model, language);
        if (text) {
          messageText = messageText ? `${messageText}\n\n${text}` : text;
          logger.info(`[telegram] Transcribed ${text.length} chars`);
        } else {
          logger.warn("[telegram] Transcription returned empty text");
        }
      } finally {
        try { fs.rmSync(tmpDir, { recursive: true, force: true }); } catch {}
      }
    }
  } catch (err) {
    logger.error(
      `[telegram] STT failed: ${err instanceof Error ? err.message : err}`,
    );
    try {
      await this.bot.sendMessage(
        telegramMsg.chat.id,
        "⚠️ Couldn't transcribe your voice message. Please try again or type instead.",
      );
    } catch {}
    return; // don't forward an empty message downstream
  }
}

// then use `messageText` in place of `telegramMsg.text || ""` when constructing `msg`
```

### Related: end-to-end setup from the gateway only

The portal's "Download Whisper" flow is misleading today: it downloads the model, but the user still has to leave the gateway, install `whisper-cli` (build from source) and `ffmpeg` manually, and put them on the gateway's PATH. Without these, transcription fails silently with `ENOENT: whisper-cli` and the user is left wondering why their voice messages don't work.

To make voice messages truly one-click, the gateway should be able to provision the runtime deps the same way it provisions the model. Some options, from simplest to most complete:

1. **Status surfacing.** `GET /api/stt/status` already exists for the model — extend it to report binary presence (`whisperCli: "found" | "missing"`, `ffmpeg: "found" | "missing"`) and surface that in the portal. Today the portal happily says "ready" while the binary is missing.
2. **Setup wizard with copy-paste install commands.** When binaries are missing, the portal shows a per-OS snippet (e.g. `brew install ffmpeg whisper-cpp` on macOS, build steps for Linux). Low engineering cost, still requires the user to run shell commands.
3. **Managed bundled binaries.** Mirror the model-download pattern: the gateway downloads pre-built static binaries for the platform (whisper.cpp ships releases; ffmpeg has reliable static builds at johnvansickle.com/ffmpeg) into a managed directory (e.g. `~/.jinn/bin/`), and the STT module resolves `WHISPER_CLI` / `FFMPEG` from there before falling back to PATH. This makes voice messages work with zero terminal time.

I'd vote for #3 long-term and #1 as an immediate win — the current "model downloaded but binaries missing" state is the worst UX (looks like it should work, doesn't).

For reference, the manual install I had to do on Amazon Linux 2023 to get past the missing binaries:

```bash
# ffmpeg (static build)
curl -sL https://johnvansickle.com/ffmpeg/releases/ffmpeg-release-amd64-static.tar.xz \
  | tar -xJ -C /tmp
sudo cp /tmp/ffmpeg-*-amd64-static/{ffmpeg,ffprobe} /usr/local/bin/

# whisper-cli (build from source)
sudo dnf install -y git cmake gcc-c++ make
git clone --depth=1 https://github.com/ggml-org/whisper.cpp /tmp/whisper.cpp
cd /tmp/whisper.cpp
cmake -B build -DCMAKE_BUILD_TYPE=Release && cmake --build build -j --target whisper-cli
sudo cp build/bin/whisper-cli /usr/local/bin/
sudo cp build/src/libwhisper.so* build/ggml/src/libggml*.so* /usr/local/lib/
sudo ldconfig

# ensure on the gateway's PATH (default PATH lacks /usr/local/bin)
sudo ln -sf /usr/local/bin/{ffmpeg,whisper-cli} /usr/bin/
```

A user who can do all that probably doesn't need a managed gateway in the first place.

### Notes / open questions

- I prototyped this against the published `0.10.0` dist on my own instance and it works end-to-end (Whisper `small`, EN+PT, ~30s CPU per 6s audio on a t3-class box).
- Auto language detection (`-l auto`) is used when `languages` has more than one entry; falls back to the single configured language otherwise. This matches the existing `resolveLanguages` helper.
- On STT failure the connector now drops the message and notifies the user instead of forwarding an empty message — this also fixes the "No deferred tool marker" crash because the engine never receives an empty resume.
- Open question: should other connectors (Slack, Discord, WhatsApp) get the same treatment? Voice-message support is most natural in Telegram and WhatsApp; Slack/Discord use `audio` attachments differently.
- Happy to send a PR if the approach looks right.

### Environment

- `jinn-cli@0.10.0`
- Node 24.15.0
- Amazon Linux 2023
- whisper.cpp (built from source), ffmpeg static 7.0.2

--- a/packages/jimmy/src/connectors/telegram/index.ts
+++ b/packages/jimmy/src/connectors/telegram/index.ts
@@ -40,6 +40,7 @@ export class TelegramConnector implements Connector {
 
   private readonly sttConfig?: TelegramConnectorConfig["stt"];
   private sttChain: Promise<unknown> = Promise.resolve();
+  private sttPending = 0;
 
   constructor(config: TelegramConnectorConfig) {
     this.bot = new TelegramBot(config.botToken, { polling: false });
@@ -139,6 +140,20 @@ export class TelegramConnector implements Connector {
         const language = langs.length === 1 ? langs[0] : "auto";
 
         // Serialize transcriptions: parallel whisper-cli runs OOM on small hosts.
+        // If another transcription is already in flight, send a one-line ack so
+        // the user doesn't sit through ~duration × queue position in silence.
+        this.sttPending++;
+        if (this.sttPending > 1) {
+          try {
+            await this.bot.sendMessage(
+              telegramMsg.chat.id,
+              "⏳ Transcribing a previous voice message — yours is queued.",
+            );
+          } catch {
+            /* non-fatal */
+          }
+        }
+
         const myTurn = this.sttChain.then(async () => {
           try {
             await this.bot.sendChatAction(telegramMsg.chat.id, "typing");
@@ -163,7 +178,11 @@ export class TelegramConnector implements Connector {
             }
           }
         });
-        this.sttChain = myTurn.catch(() => undefined);
+        this.sttChain = myTurn
+          .catch(() => undefined)
+          .finally(() => {
+            this.sttPending = Math.max(0, this.sttPending - 1);
+          });
 
         let transcript: string | undefined;
         try {

--- a/packages/jimmy/src/connectors/telegram/index.ts
+++ b/packages/jimmy/src/connectors/telegram/index.ts
@@ -40,7 +40,6 @@ export class TelegramConnector implements Connector {
 
   private readonly sttConfig?: TelegramConnectorConfig["stt"];
   private sttChain: Promise<unknown> = Promise.resolve();
-  private sttQueueDepth = 0;
 
   constructor(config: TelegramConnectorConfig) {
     this.bot = new TelegramBot(config.botToken, { polling: false });
@@ -105,91 +104,101 @@ export class TelegramConnector implements Connector {
       let messageText: string =
         (telegramMsg as any).text || (telegramMsg as any).caption || "";
 
-      // Voice / audio / video_note → transcribe via STT module
+      // Voice / audio / video_note → transcribe via STT module.
+      // If STT can't run for any reason, drop the message with a user-facing
+      // explanation rather than forwarding empty text downstream (which would
+      // crash session resume — see #54).
       const voiceLike =
         (telegramMsg as any).voice ||
         (telegramMsg as any).audio ||
         (telegramMsg as any).video_note;
 
-      if (voiceLike && this.sttConfig?.enabled) {
-        const model = this.sttConfig.model || "small";
-        if (!getModelPath(model)) {
-          logger.warn(
-            `[telegram] STT model '${model}' not downloaded — skipping transcription`,
-          );
-        } else {
-          const langs = resolveLanguages(this.sttConfig);
-          const language = langs.length === 1 ? langs[0] : "auto";
+      if (voiceLike) {
+        const model = this.sttConfig?.model || "small";
+        let unavailable: string | null = null;
+        if (!this.sttConfig?.enabled) {
+          unavailable = "voice transcription is not enabled on this gateway";
+        } else if (!getModelPath(model)) {
+          unavailable = `STT model '${model}' is not downloaded`;
+        }
 
-          // Serialize transcriptions: parallel whisper-cli runs OOM on small boxes.
-          this.sttQueueDepth = (this.sttQueueDepth || 0) + 1;
-          const myPosition = this.sttQueueDepth;
-          if (myPosition > 1) {
-            logger.info(
-              `[telegram] Voice message queued (position ${myPosition}, ${myPosition - 1} ahead)`,
-            );
-            try {
-              await this.bot.sendMessage(
-                telegramMsg.chat.id,
-                `⏳ Queued — ${myPosition - 1} voice message${myPosition - 1 === 1 ? "" : "s"} ahead. I'll transcribe yours when its turn comes up.`,
-              );
-            } catch {
-              /* non-fatal */
-            }
-          }
-
-          const myTurn = this.sttChain.then(async () => {
-            try {
-              await this.bot.sendChatAction(telegramMsg.chat.id, "typing");
-            } catch {
-              /* non-fatal */
-            }
-            logger.info(
-              `[telegram] Transcribing voice message (${voiceLike.duration}s, lang=${language})`,
-            );
-            const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "tg-stt-"));
-            try {
-              const localPath = await (this.bot as any).downloadFile(
-                voiceLike.file_id,
-                tmpDir,
-              );
-              return await sttTranscribe(localPath, model, language);
-            } finally {
-              try {
-                fs.rmSync(tmpDir, { recursive: true, force: true });
-              } catch {
-                /* non-fatal */
-              }
-            }
-          });
-          this.sttChain = myTurn
-            .catch(() => undefined)
-            .finally(() => {
-              this.sttQueueDepth = Math.max(0, (this.sttQueueDepth || 1) - 1);
-            });
-
+        if (unavailable) {
+          logger.warn(`[telegram] Dropping voice message: ${unavailable}`);
           try {
-            const text = await myTurn;
-            if (text) {
-              messageText = messageText ? `${messageText}\n\n${text}` : text;
-              logger.info(`[telegram] Transcribed ${text.length} chars`);
-            } else {
-              logger.warn("[telegram] Transcription returned empty text");
-            }
-          } catch (err) {
-            logger.error(
-              `[telegram] STT failed: ${err instanceof Error ? err.message : err}`,
+            await this.bot.sendMessage(
+              telegramMsg.chat.id,
+              `⚠️ Couldn't transcribe your voice message — ${unavailable}. Please type instead.`,
             );
+          } catch {
+            /* non-fatal */
+          }
+          return;
+        }
+
+        const langs = resolveLanguages(this.sttConfig);
+        const language = langs.length === 1 ? langs[0] : "auto";
+
+        // Serialize transcriptions: parallel whisper-cli runs OOM on small hosts.
+        const myTurn = this.sttChain.then(async () => {
+          try {
+            await this.bot.sendChatAction(telegramMsg.chat.id, "typing");
+          } catch {
+            /* non-fatal */
+          }
+          logger.info(
+            `[telegram] Transcribing voice message (${voiceLike.duration}s, lang=${language})`,
+          );
+          const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "tg-stt-"));
+          try {
+            const localPath = await (this.bot as any).downloadFile(
+              voiceLike.file_id,
+              tmpDir,
+            );
+            return await sttTranscribe(localPath, model, language);
+          } finally {
             try {
-              await this.bot.sendMessage(
-                telegramMsg.chat.id,
-                "⚠️ Couldn't transcribe your voice message. Please try again or type instead.",
-              );
+              fs.rmSync(tmpDir, { recursive: true, force: true });
             } catch {
               /* non-fatal */
             }
-            return;
           }
+        });
+        this.sttChain = myTurn.catch(() => undefined);
+
+        let transcript: string | undefined;
+        try {
+          transcript = await myTurn;
+        } catch (err) {
+          logger.error(
+            `[telegram] STT failed: ${err instanceof Error ? err.message : err}`,
+          );
+          try {
+            await this.bot.sendMessage(
+              telegramMsg.chat.id,
+              "⚠️ Couldn't transcribe your voice message. Please try again or type instead.",
+            );
+          } catch {
+            /* non-fatal */
+          }
+          return;
+        }
+
+        if (transcript) {
+          messageText = messageText
+            ? `${messageText}\n\n${transcript}`
+            : transcript;
+          logger.info(`[telegram] Transcribed ${transcript.length} chars`);
+        } else {
+          logger.warn("[telegram] Transcription returned empty text");
+          try {
+            await this.bot.sendMessage(
+              telegramMsg.chat.id,
+              "⚠️ Couldn't make out anything in your voice message. Please try again or type instead.",
+            );
+          } catch {
+            /* non-fatal */
+          }
+          return;
         }
       }
 

--- a/packages/jimmy/src/connectors/telegram/index.ts
+++ b/packages/jimmy/src/connectors/telegram/index.ts
@@ -1,4 +1,7 @@
 import TelegramBot from "node-telegram-bot-api";
+import fs from "node:fs";
+import os from "node:os";
+import path from "node:path";
 import type {
   Connector,
   ConnectorCapabilities,
@@ -11,6 +14,11 @@ import type {
 import { deriveSessionKey, buildReplyContext, isOldTelegramMessage } from "./threads.js";
 import { formatResponse } from "./format.js";
 import { logger } from "../../shared/logger.js";
+import {
+  transcribe as sttTranscribe,
+  resolveLanguages,
+  getModelPath,
+} from "../../stt/stt.js";
 
 export class TelegramConnector implements Connector {
   name = "telegram";
@@ -30,6 +38,10 @@ export class TelegramConnector implements Connector {
     attachments: true,
   };
 
+  private readonly sttConfig?: TelegramConnectorConfig["stt"];
+  private sttChain: Promise<unknown> = Promise.resolve();
+  private sttQueueDepth = 0;
+
   constructor(config: TelegramConnectorConfig) {
     this.bot = new TelegramBot(config.botToken, { polling: false });
     this.ignoreOldMessagesOnBoot = config.ignoreOldMessagesOnBoot !== false;
@@ -37,6 +49,7 @@ export class TelegramConnector implements Connector {
       config.allowFrom && config.allowFrom.length > 0
         ? new Set(config.allowFrom)
         : null;
+    this.sttConfig = config.stt;
   }
 
   async start(): Promise<void> {
@@ -89,6 +102,97 @@ export class TelegramConnector implements Connector {
       const username =
         telegramMsg.from?.username || telegramMsg.from?.first_name || "unknown";
 
+      let messageText: string =
+        (telegramMsg as any).text || (telegramMsg as any).caption || "";
+
+      // Voice / audio / video_note → transcribe via STT module
+      const voiceLike =
+        (telegramMsg as any).voice ||
+        (telegramMsg as any).audio ||
+        (telegramMsg as any).video_note;
+
+      if (voiceLike && this.sttConfig?.enabled) {
+        const model = this.sttConfig.model || "small";
+        if (!getModelPath(model)) {
+          logger.warn(
+            `[telegram] STT model '${model}' not downloaded — skipping transcription`,
+          );
+        } else {
+          const langs = resolveLanguages(this.sttConfig);
+          const language = langs.length === 1 ? langs[0] : "auto";
+
+          // Serialize transcriptions: parallel whisper-cli runs OOM on small boxes.
+          this.sttQueueDepth = (this.sttQueueDepth || 0) + 1;
+          const myPosition = this.sttQueueDepth;
+          if (myPosition > 1) {
+            logger.info(
+              `[telegram] Voice message queued (position ${myPosition}, ${myPosition - 1} ahead)`,
+            );
+            try {
+              await this.bot.sendMessage(
+                telegramMsg.chat.id,
+                `⏳ Queued — ${myPosition - 1} voice message${myPosition - 1 === 1 ? "" : "s"} ahead. I'll transcribe yours when its turn comes up.`,
+              );
+            } catch {
+              /* non-fatal */
+            }
+          }
+
+          const myTurn = this.sttChain.then(async () => {
+            try {
+              await this.bot.sendChatAction(telegramMsg.chat.id, "typing");
+            } catch {
+              /* non-fatal */
+            }
+            logger.info(
+              `[telegram] Transcribing voice message (${voiceLike.duration}s, lang=${language})`,
+            );
+            const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "tg-stt-"));
+            try {
+              const localPath = await (this.bot as any).downloadFile(
+                voiceLike.file_id,
+                tmpDir,
+              );
+              return await sttTranscribe(localPath, model, language);
+            } finally {
+              try {
+                fs.rmSync(tmpDir, { recursive: true, force: true });
+              } catch {
+                /* non-fatal */
+              }
+            }
+          });
+          this.sttChain = myTurn
+            .catch(() => undefined)
+            .finally(() => {
+              this.sttQueueDepth = Math.max(0, (this.sttQueueDepth || 1) - 1);
+            });
+
+          try {
+            const text = await myTurn;
+            if (text) {
+              messageText = messageText ? `${messageText}\n\n${text}` : text;
+              logger.info(`[telegram] Transcribed ${text.length} chars`);
+            } else {
+              logger.warn("[telegram] Transcription returned empty text");
+            }
+          } catch (err) {
+            logger.error(
+              `[telegram] STT failed: ${err instanceof Error ? err.message : err}`,
+            );
+            try {
+              await this.bot.sendMessage(
+                telegramMsg.chat.id,
+                "⚠️ Couldn't transcribe your voice message. Please try again or type instead.",
+              );
+            } catch {
+              /* non-fatal */
+            }
+            return;
+          }
+        }
+      }
+
       const msg: IncomingMessage = {
         connector: this.name,
         source: "telegram",
@@ -98,7 +202,7 @@ export class TelegramConnector implements Connector {
         channel: String(telegramMsg.chat.id),
         user: username,
         userId: String(userId ?? "unknown"),
-        text: telegramMsg.text || "",
+        text: messageText,
         attachments: [],
         raw: telegramMsg,
         transportMeta: {

--- a/packages/jimmy/src/gateway/server.ts
+++ b/packages/jimmy/src/gateway/server.ts
@@ -364,6 +364,7 @@ export async function startGateway(
         botToken: config.connectors.telegram.botToken,
         allowFrom: config.connectors.telegram.allowFrom,
         ignoreOldMessagesOnBoot: config.connectors.telegram.ignoreOldMessagesOnBoot,
+        stt: config.stt,
       });
       telegram.onMessage((msg) => {
         const routeOpts: RouteOptions = {};
@@ -480,7 +481,7 @@ export async function startGateway(
             break;
           }
           case "telegram": {
-            const telegramConfig = { ...typeConfig, id } as any;
+            const telegramConfig = { ...typeConfig, id, stt: config.stt } as any;
             const tg = new TelegramConnector(telegramConfig);
             tg.onMessage((msg) => {
               const routeOpts: RouteOptions = {};
@@ -605,7 +606,7 @@ export async function startGateway(
               break;
             }
             case "telegram": {
-              const telegramConfig = { ...typeConfig, id } as any;
+              const telegramConfig = { ...typeConfig, id, stt: config.stt } as any;
               const tg = new TelegramConnector(telegramConfig);
               tg.onMessage((msg) => {
                 const routeOpts: RouteOptions = {};

--- a/packages/jimmy/src/shared/types.ts
+++ b/packages/jimmy/src/shared/types.ts
@@ -330,6 +330,13 @@ export interface TelegramConnectorConfig {
   botToken: string;
   allowFrom?: number[];
   ignoreOldMessagesOnBoot?: boolean;
+  /** Speech-to-text settings forwarded from top-level `config.stt` */
+  stt?: {
+    enabled?: boolean;
+    model?: string;
+    language?: string;
+    languages?: string[];
+  };
 }
 
 export interface WhatsAppConnectorConfig {

